### ISSUE TITLE
Incorrectly assumes EOL if connection fails

### ISSUE THREAD

**Issue by tfogwill** [xolox/python-apt-mirror-updater#1]
If a connection to the security repository (e.g. security.ubuntu.com) fails,  AptMirrorUpdater.validate_mirror fails, causing AptMirrorUpdater.release_is_eol to return true and overwriting the possibly valid mirror with old_releases_url (e.g. old-releases.ubuntu.com)

**Comment by xolox** [comment]
Hi @tfogwill and thanks for the bug report here and the pull request in #2. The issue you reported should now be fixed, in more than one way:

1. Initially I merged your pull request without changes.
2. Then I had the insight that the bug you reported here was caused by my logic in `release_is_eol` which should have been checking specifically for an HTTP response with a 404 status code but instead incorrectly assumed that any error condition implied EOL.
3. Triggered by point 2 I refactored the code base to distinguish HTTP 404 responses (which can indicate EOL) from other error conditions (like connection and timeout errors).
4. The changes from point 3 obviated the need for the additional HTTP request performed by `can_connect_to_mirror()` so I removed it and changed `release_is_eol` to check for the new `MirrorStatus.MAYBE_EOL` return value.
5. Finally I implemented more robust and deterministic EOL detection in 823f7587f4f2885a43270b5e1eca32932d7a7fb1 based on a data structure with known EOL dates of Debian and Ubuntu releases. The `MirrorStatus.MAYBE_EOL` logic now remains only as a fall back for unknown releases.

These changes have been published to PyPI as version 5.0.
